pkg/concurrency/runner.go

```diff
@@ -0,0 +1,89 @@
+package concurrency
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+	"sync"
+)
+
+// Runner is a function that runs a task.
+type Runner func(context.Context) error
+
+// RunnerManager is a manager for runners. It runs all runners in parallel and
+// waits for all runners to finish. If any runner returns, the RunnerManager
+// will stop all other runners and return any error.
+type RunnerManager struct {
+	errs    []string
+	lock    sync.Mutex
+	wg      sync.WaitGroup
+	runners []Runner
+	running chan struct{}
+}
+
+// NewRunnerManager creates a new RunnerManager.
+func NewRunnerManager(runners ...Runner) *RunnerManager {
+	return &RunnerManager{
+		runners: runners,
+		running: make(chan struct{}),
+	}
+}
+
+// Add adds a new runner to the RunnerManager.
+func (r *RunnerManager) Add(runner Runner) error {
+	select {
+	case <-r.running:
+		return errors.New("manager already started")
+	default:
+	}
+
+	r.lock.Lock()
+	defer r.lock.Unlock()
+	r.runners = append(r.runners, runner)
+	return nil
+}
+
+// Run runs all runners in parallel and waits for all runners to finish. If any
+// runner returns, the RunnerManager will stop all other runners and return any
+// error.
+func (r *RunnerManager) Run(ctx context.Context) error {
+	select {
+	case <-r.running:
+		return errors.New("manager already started")
```

Comment:
could you please extract this and reuse as a variable?

`var ErrManagerAlreadyStarted = errors.New("manager already started")